qiskit/pulse/schedule.py

```diff
@@ -1180,38 +1434,60 @@ def replace(
         Returns:
             The modified schedule block with ``old`` replaced by ``new``.
         """
-        from qiskit.pulse.parameter_manager import ParameterManager
-
-        new_blocks = []
-        new_parameters = ParameterManager()
-
-        for block in self.blocks:
-            if block == old:
-                new_blocks.append(new)
-                new_parameters.update_parameter_table(new)
-            else:
-                if isinstance(block, ScheduleBlock):
-                    new_blocks.append(block.replace(old, new, inplace))
-                else:
-                    new_blocks.append(block)
-                new_parameters.update_parameter_table(block)
+        if not inplace:
+            schedule = copy.deepcopy(self)
+            return schedule.replace(old, new, inplace=True)
+
+        # Temporarily copies references
+        all_references = ReferenceManager()
+        all_references.update(self.references)
+
+        if isinstance(new, ScheduleBlock):
+            new = copy.deepcopy(new)
+
+            # This dict.update bypasses existing reference check.
+            # Overriding existing reference should be okey because we replace the block.
+            all_references.update(new.references)
+            new._reference_manager.clear()
+            new._parent = self
+
+        # Regenerate parameter table by regenerating elements.
+        # Note that removal of parameters in old is not sufficient,
+        # because corresponding parameters might be also used in another block element.
+        self._parameter_manager.clear()
+        self._parameter_manager.update_parameter_table(self._alignment_context)
+        new_elms = []
+        for elm in self._blocks:
+            if elm == old:
+                elm = new
+            self._parameter_manager.update_parameter_table(elm)
+            new_elms.append(elm)
```

Comment:
Can the same element be in the blocks more than once?  If so, it could get its parameter table updated twice which may cause an error?  If not, you can probably short-circuit the loop once you've seen the target to avoid what I assume are costly `==` comparisons.

Reply:
Yes, it can. Note that we cannot shortcut the loop because we need to refresh parameter manager, i.e. we could think of simply removing parameters associated with `old`, however, the same parameter could exist in other block elements. Thus we always need full scan. I think replace is not good method and usually it's better to write a schedule from scratch.

Reply:
Fine by me, as long as it was intentional.